src/SAML2/Assertion.php

```diff
@@ -1654,91 +1608,8 @@ private function addAttributeStatement(DOMElement $root): void
-                        /** @var \DOMElement $value */
-                        $value = $eptiValue->item(0);
-                        $node = $root->ownerDocument->importNode($value, true);
-                        $attributeValue->appendChild($node);
-                    } else {
-                        $attributeValue->textContent = $eptiValue;
-                    }
-                }
-
-                continue;
-            }
-
-            // get value type(s) for the current attribute
-            if (array_key_exists($name, $this->attributesValueTypes)) {
-                $valueTypes = $this->attributesValueTypes[$name];
-                if (is_array($valueTypes) && count($valueTypes) != count($values)) {
-                    throw new \Exception('Array of value types and array of values have different size for attribute ' .
-                        var_export($name, true));
-                }
-            } else {
-                // if no type(s), default behaviour
-                $valueTypes = null;
-            }
-
-            $vidx = -1;
-            foreach ($values as $value) {
-                $vidx++;
-
-                // try to get type from current types
-                $type = null;
-                if (!is_null($valueTypes)) {
-                    if (is_array($valueTypes)) {
-                        $type = $valueTypes[$vidx];
-                    } else {
-                        $type = $valueTypes;
-                    }
-                }
-
-                // if no type get from types, use default behaviour
-                if (is_null($type)) {
-                    if (is_string($value)) {
-                        $type = 'xs:string';
-                    } elseif (is_int($value)) {
-                        $type = 'xs:integer';
-                    } else {
-                        $type = null;
-                    }
-                }
-
-                $attributeValue = $document->createElementNS(Constants::NS_SAML, 'saml:AttributeValue');
-                $attribute->appendChild($attributeValue);
-                if ($type !== null) {
-                    $attributeValue->setAttributeNS(Constants::NS_XSI, 'xsi:type', $type);
-                }
-                if (is_null($value)) {
-                    $attributeValue->setAttributeNS(Constants::NS_XSI, 'xsi:nil', 'true');
-                }
-
-                if ($value instanceof \DOMNodeList) {
-                    foreach ($value as $v) {
-                        $node = $document->importNode($v, true);
-                        $attributeValue->appendChild($node);
-                    }
-                } else {
-                    $value = strval($value);
-                    $attributeValue->appendChild($document->createTextNode($value));
-                }
-            }
+        foreach ($this->attributes as $name => $attributeObj) {
+            $attributeObj->toXML($attributeStatement);
```

Comment:
I think this is really great because it allows us to offload the responsibility to know how to marshall attributes to the attributes themselves, instead of needing code handling exceptional cases here. However, I'm not entirely sure it would work the way it is. EPTID should contain a `NameID` in its value, but I don't think that¡s the way it is implemented. Could it be an option to have a class representing this attribute in particular? It would just extend the `Attribute` class. However, we would then need a factory to create attributes.

Reply:
Oh, nevermind, I just saw the modifications to `AttributeValue`... 🤦‍♂